arch/Xtensa/XtensaDisassembler.c

```diff
@@ -0,0 +1,302 @@
+/* https://github.com/capstone-engine/capstone/tree/next/suite/auto-sync */
+
+//===-- XtensaDisassembler.cpp - Disassembler for Xtensa ------------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the XtensaDisassembler class.
+//
+//===----------------------------------------------------------------------===//
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <capstone/platform.h>
+
+#include "../../MathExtras.h"
+#include "../../MCDisassembler.h"
+#include "../../MCFixedLenDisassembler.h"
+#include "../../SStream.h"
+#include "../../cs_priv.h"
+#include "priv.h"
+
+#define GET_INSTRINFO_MC_DESC
+#include "XtensaGenInstrInfo.inc"
+
+#define CONCAT(a, b) CONCAT_(a, b)
+#define CONCAT_(a, b) a##_##b
+
+#define DEBUG_TYPE "Xtensa-disassembler"
+
+static const unsigned ARDecoderTable[] = {
+	Xtensa_A0,  Xtensa_SP,	Xtensa_A2,  Xtensa_A3, Xtensa_A4,  Xtensa_A5,
+	Xtensa_A6,  Xtensa_A7,	Xtensa_A8,  Xtensa_A9, Xtensa_A10, Xtensa_A11,
+	Xtensa_A12, Xtensa_A13, Xtensa_A14, Xtensa_A15
+};
+
+static DecodeStatus DecodeARRegisterClass(MCInst *Inst, uint64_t RegNo,
+					  uint64_t Address, const void *Decoder)
+{
+	if (RegNo >= sizeof(ARDecoderTable))
+		return MCDisassembler_Fail;
+
+	unsigned Reg = ARDecoderTable[RegNo];
+	MCOperand_CreateReg0(Inst, (Reg));
+	return MCDisassembler_Success;
+}
+
+static const unsigned SRDecoderTable[] = { Xtensa_SAR, 3 };
+
+static DecodeStatus DecodeSRRegisterClass(MCInst *Inst, uint64_t RegNo,
+					  uint64_t Address, const void *Decoder)
+{
+	if (RegNo > 255)
+		return MCDisassembler_Fail;
+
+	for (unsigned i = 0; i + 1 < sizeof(SRDecoderTable); i += 2) {
+		if (SRDecoderTable[i + 1] == RegNo) {
+			unsigned Reg = SRDecoderTable[i];
+			MCOperand_CreateReg0(Inst, (Reg));
+			return MCDisassembler_Success;
+		}
+	}
+
+	return MCDisassembler_Fail;
+}
+
+static bool tryAddingSymbolicOperand(int64_t Value, bool isBranch,
+				     uint64_t Address, uint64_t Offset,
+				     uint64_t InstSize, MCInst *MI,
+				     const void *Decoder)
+{
+	//	return Dis->tryAddingSymbolicOperand(MI, Value, Address, isBranch,
+	//					     Offset, /*OpSize=*/0, InstSize);
+	return true;
+}
+
+static DecodeStatus decodeCallOperand(MCInst *Inst, uint64_t Imm,
+				      int64_t Address, const void *Decoder)
+{
+	CS_ASSERT(CONCAT(isUInt, 18)(Imm) && "Invalid immediate");
+	MCOperand_CreateImm0(Inst, (SignExtend64((Imm << 2), 20)));
+	return MCDisassembler_Success;
+}
+
+static DecodeStatus decodeJumpOperand(MCInst *Inst, uint64_t Imm,
+				      int64_t Address, const void *Decoder)
+{
+	CS_ASSERT(CONCAT(isUInt, 18)(Imm) && "Invalid immediate");
```

Comment:
```suggestion
	CS_ASSERT(isUIntN(18, Imm) && "Invalid immediate");
```

This function is provided in `MathExtras.h`. I added the patch for it also in the SystemZ PR.

Reply:
Please also use it everywhere else.

Reply:
I just see, that the documentation of the `isUInt` Patch is off:
```python
class IsUInt(Patch):
    """
    Patch   isUInt<N>(...)
    to      isUInt(..., N)  # Should be: "isUIntN(N, ...)"
    """
```

If you want to, you can fix it.